dask/array/routines.py

```diff
@@ -312,25 +312,56 @@ def _inner_apply_along_axis(arr,
     )
 
 
-@wraps(np.apply_along_axis)
-def apply_along_axis(func1d, axis, arr, *args, **kwargs):
+def apply_along_axis(func1d, axis, arr, dtype=None, shape=None, *args, **kwargs):
+    """
+    Apply a function to 1-D slices along the given axis. This is
+    a blocked variant of :func:`numpy.apply_along_axis` implemented via
+    :func:`dask.array.map_blocks`
+
+    Parameters
+    __________
+
+    func1d : callable
+        Function to apply to 1-D slices of the array along the given axis
+    axis : int
+        Axis along which func1d will be applied
+    arr : dask array
+        Dask array to which ``func1d`` will be applied
+    dtype : np.dtype, optional
+        The ``dtype`` of the output array. It is recommended to provide this.
+        If not provided, will be inferred by applying the function to a small
+        set of fake data.
+    shape : tuple, optional
+        The ``shape`` of the output array. It is recommended to provide this.
+        If not provided, will be inferred by applying the function to a small
+        set of fake data.
+    args : any
+        Additional arguments to ``func1d``.
+    kwargs : any
+        Additional named arguments to ``func1d``.
+
+    """
     arr = asarray(arr)
 
     # Validate and normalize axis.
-    arr.shape[axis]
     axis = len(arr.shape[:axis])
 
     # Test out some data with the function.
-    test_data = np.ones((1,), dtype=arr.dtype)
-    test_result = np.array(func1d(test_data, *args, **kwargs))
+    if (shape is None) or (dtype is None):
+        test_data = np.ones((1,), dtype=arr.dtype)
+        test_result = np.array(func1d(test_data, *args, **kwargs))
+        if shape is None:
+            shape = test_result.shape
+        if dtype is None:
+            dtype = test_result.dtype
```

Comment:
So we might want to restructure this a bit. Namely if `shape` and `dtype` are both specified, we need not test at all. If `shape` is specified, guess we need to pass a chunk sized array in? Not sure. Feel free to share thoughts. Specifying `dtype` alone doesn't really bypass the check if we need `shape`. The original test can be run if neither is specified.

Reply:
if `dtype` and `shape` are both specified, the test [will not run](https://github.com/d-v-b/dask/blob/master/dask/array/routines.py#L350). 

For this PR I thought it made sense to use the old test code for cases when either `shape` or `dtype` are not specified. I thought doing anything fancier with `dtype` / `shape` inference should be in a separate PR that potentially leverages [dask.array.core.apply_infer_dtype](https://github.com/dask/dask/blob/master/dask/array/core.py#L553)